pkg/handlers/generic/mutation/encryption/inject.go

```diff
@@ -0,0 +1,307 @@
+	providers []carenv1.EncryptionProviders,
+) (*apiserverv1.EncryptionConfiguration, error) {
+	resourceConfigs := []apiserverv1.ResourceConfiguration{}
+	for _, encProvider := range providers {
+		provider := encProvider
+		resourceConfig, err := encryptionConfigForSecretsAndConfigMaps(
+			&provider,
+			h.keyGenerator,
+		)
+		if err != nil {
+			return nil, err
+		}
+		resourceConfigs = append(resourceConfigs, *resourceConfig)
+	}
+	// We only support encryption for "secrets" and "configmaps" using "aescbc" provider.
+
+	return &apiserverv1.EncryptionConfiguration{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: apiserverv1.SchemeGroupVersion.String(),
+			Kind:       "EncryptionConfiguration",
+		},
+		Resources: resourceConfigs,
+	}, nil
+}
+
+func (h *encryptionPatchHandler) DefaultEncryptionSecretExists(
+	ctx context.Context,
+	cluster *clusterv1.Cluster,
+) (bool, error) {
+	secretName := defaultEncryptionSecretName(cluster.Name)
+	existingSecret := &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: corev1.SchemeGroupVersion.String(),
+			Kind:       "Secret",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: cluster.Namespace,
+		},
+	}
+	err := h.client.Get(ctx, ctrlclient.ObjectKeyFromObject(existingSecret), existingSecret)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func (h *encryptionPatchHandler) CreateEncryptionConfigurationSecret(
+	ctx context.Context,
+	encryptionConfig *apiserverv1.EncryptionConfiguration,
+	cluster *clusterv1.Cluster,
+) error {
+	dataYaml, err := yaml.Marshal(encryptionConfig)
+	if err != nil {
+		return fmt.Errorf("unable to marshal encryption configuration to YAML: %w", err)
+	}
+
+	secretData := map[string]string{
+		SecretKeyForEtcdEncryption: strings.TrimSpace(string(dataYaml)),
+	}
+	secretName := defaultEncryptionSecretName(cluster.Name)
+	encryptionConfigSecret := &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: corev1.SchemeGroupVersion.String(),
+			Kind:       "Secret",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: cluster.Namespace,
+			Labels:    utils.NewLabels(utils.WithMove(), utils.WithClusterName(cluster.Name)),
+		},
+		StringData: secretData,
+		Type:       corev1.SecretTypeOpaque,
+	}
+
+	if err = controllerutil.SetOwnerReference(cluster, encryptionConfigSecret, h.client.Scheme()); err != nil {
+		return fmt.Errorf(
+			"failed to set owner reference on encryption configuration secret: %w",
+			err,
+		)
+	}
+
+	// We only support creating encryption config in BeforeClusterCreate hook and ensure that the keys are immutable.
+	if err := k8sClientUtil.Create(ctx, h.client, encryptionConfigSecret); err != nil {
+		return fmt.Errorf("failed to create encryption configuration secret: %w", err)
+	}
+	return nil
+}
+
+// We only support encryption for "secrets" and "configmaps".
+func encryptionConfigForSecretsAndConfigMaps(
```

Comment:
nit: maybe something a little more generic so it doesn't need to change if the details change?
```suggestion
func defaultEncryptionConfiguration(
```

Reply:
fixed https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/commit/89d1408e8533448aa0168d904a7add4896d6150b